src/internal/coredb/projects.go

```diff
@@ -14,33 +15,70 @@ import (
 	"github.com/pachyderm/pachyderm/v2/src/pfs"
 )
 
+// ProjectIterator batches a page of projectRow entries. Entries can be retrieved using iter.Next().
 type ProjectIterator struct {
-	*sqlx.Rows
+	lock     sync.Mutex
+	pageNum  int
+	projects []projectRow
+	index    int
+	db       *pachsql.DB
 }
 
-// Next advances the iterator by one row.
-func (iter *ProjectIterator) Next() (*pfs.ProjectInfo, error) {
-	project := &pfs.ProjectInfo{Project: &pfs.Project{}}
-	var createdAt time.Time
-	if iter.Rows.Next() {
-		if err := iter.Rows.Scan(&project.Project.Name, &project.Description, &createdAt); err != nil {
-			return nil, errors.Wrap(err, "failed to scan row")
-		}
-		projectTimestamp, err := types.TimestampProto(createdAt)
+type projectRow struct {
+	Name        string    `db:"name"`
+	Description string    `db:"description"`
+	CreatedAt   time.Time `db:"created_at"`
+}
+
+// Next advances the iterator by one row. It returns an io.EOF when there are no more entries.
+func (iter *ProjectIterator) Next(ctx context.Context, project *pfs.ProjectInfo) error {
+	if project == nil {
+		return errors.Wrap(fmt.Errorf("project is nil"), "failed to get next project")
+	}
+	var err error
+	iter.lock.Lock()
+	if iter.index >= len(iter.projects) {
+		iter.index = 0
+		iter.pageNum++
+		iter.projects, err = listProjectPage(ctx, iter.db, 100, iter.pageNum)
 		if err != nil {
-			return nil, errors.Wrap(err, "failed to convert time.Time to proto timestamp")
+			return errors.Wrap(err, "failed to list project page")
+		}
+		if len(iter.projects) == 0 {
+			return io.EOF
 		}
-		project.CreatedAt = projectTimestamp
-		return project, nil
 	}
-	if iter.Rows.Err() != nil {
-		return nil, errors.Wrap(iter.Rows.Err(), "failed iterating")
+	row := iter.projects[iter.index]
+	projectTimestamp, err := types.TimestampProto(row.CreatedAt)
+	if err != nil {
+		return errors.Wrap(err, "failed to convert time.Time to proto timestamp")
 	}
-	return nil, io.EOF
+	*project = pfs.ProjectInfo{
+		Project:     &pfs.Project{Name: row.Name},
+		Description: row.Description,
+		CreatedAt:   projectTimestamp}
+	iter.index++
+	iter.lock.Unlock()
+	return nil
 }
 
-func (iter *ProjectIterator) Close() error {
-	return errors.Wrap(iter.Rows.Close(), "error closing iterator")
+// ListProject returns a ProjectIterator that exposes a Next() function for retrieving *pfs.ProjectInfo references.
+func ListProject(ctx context.Context, db *pachsql.DB) (*ProjectIterator, error) {
+	page, err := listProjectPage(ctx, db, 100, 0)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list projects")
+	}
+	iter := &ProjectIterator{
+		projects: page,
+		db:       db,
+	}
+	return iter, nil
+}
+
+func listProjectPage(ctx context.Context, db *pachsql.DB, pageSize, pageNum int) ([]projectRow, error) {
+	var page []projectRow
+	err := db.SelectContext(ctx, &page, "SELECT name,description,created_at FROM core.projects ORDER BY id ASC LIMIT $1 OFFSET $2", pageSize, pageSize*pageNum)
```

Comment:
nit: I wonder if in the event of an error, `page` would be partially filled? In which case, it would be safer to do something like

```Go
	if err := db.SelectContext(ctx, &page, "SELECT name,description,created_at FROM core.projects ORDER BY id ASC LIMIT $1 OFFSET $2", pageSize, pageSize*pageNum); err != nil {
	     return nil, errors.Wrap(err, "could not to get project page")
	}
	return page, nil
```